### ISSUE TITLE
User management: passwordFile does not work as advertised on current stable NixOS

### ISSUE THREAD

**Issue by kciredor** [NixOS/nixpkgs#148044]
### Describe the bug
Using users.users.<user>.passwordFile = "path/to/textfile/containing/hashed/password"; does not work as advertised, I'm unable to login.

Using a workaround with hashedPassword sourced from the same file does work and results in a system I can login to.

### Steps To Reproduce
Bootstrapping a new system with NixOS and having a configuration.nix containing:
`
users = {
  mutableUsers = false;

  users.kciredor = {
    isNormalUser = true;
    passwordFile = "/path/to/password.txt";
  };
};
`

Followed by running `nixos-install`, results in a system I cannot log in to.

Working around the issue by replacing passwordFile with hashedPassword like this does work:
`
users = {
  mutableUsers = false;

  users.kciredor = {
    isNormalUser = true;
    hashedPassword = lib.strings.fileContents path/to/password.txt;
  };
};
`

Please note that both approaches use the exact same password input file.

Looking into https://github.com/NixOS/nixpkgs/blob/nixos-21.05/nixos/modules/config/users-groups.nix#L247 I'm confident that passwordFile should contain a hash and not a plaintext password.

I've created the hashed password using `mkpasswd -m sha-512 | tr -d '\n' > password.txt` (which also strips the trailing linebreak unfortunately produced by mkpasswd, as can be verified by a hex dump).

### Notify maintainers
Could not find any maintainers listed so I'm tagging the most recent committers to users-groups.nix:
@alyssais @symphorien, @dnr, @ncfavier, @infinisil

### Metadata
```console
$ nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/1bpw671p7r93sp3sbdxq9ly6h4npx5ff-bash-interactive-4.4-p23-dev
copying path '/nix/store/1bpw671p7r93sp3sbdxq9ly6h4npx5ff-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.81, NixOS, 21.05.4394.2553aee74fe (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"home-manager-21.05, nixos-21.05.4394.2553aee74fe, nixos-unstable-22.05pre334685.8a308775674"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
~ took 3s
```

**Comment by ncfavier** [comment]
Have you tried without `| tr -d '\n'`? That shouldn't be needed.

Can you compare the contents of `/etc/passwd` and `/etc/shadow` in both cases?

**Comment by kciredor** [comment]
Yes @ncfavier my first attempts were without the removal of newline. But because it was not working, and a hint on IRC, I took a better look at the produced hash via a hex dump, and I figured: "ah that must be it, there's a newline character added to the hash". But that was not the problem, I left it there though because for the workaround with hashedPassword it still makes sense.

To be honest /etc/shadow has a hash of '!' for my user after nixos-install. The docs state that it will be updated on system activation, which in my mind is after a boot into the new system.

So currently I'm thinking: what if nixos-install is not the same thing as nixos-rebuild? Could it be nixos-install which does not play nicely with passwordFile?

**Comment by kciredor** [comment]
As soon as I run nixos-switch rebuild from a running system, replacing hashedPassword with passwordFile, my /etc/shadow file user hash is removed and replaced by '!' again. So nixos-install and nixos-rebuild act the same way: '!' instead of the hash.

**Comment by kciredor** [comment]
Think I've got it now. As the documentation states, passwordFile must contain the full path to the file containing the password hash.

When bootstrapping a new system, everything is mounted into /mnt so I have to deal with the /mnt prefix. Using full path it means something like /mnt/path/to/password.txt.

When rebuilding a running system there is no /mnt prefix so the path would suddenly have to became /path/to/password.txt.

_For that reason I've been using a relative path in configuration.nix which works fine with other stuff like importing files, lib.strings.fileContents and so on. But not with passwordFile._

I'll stick with my workaround for this reason, sourcing the hashedPassword with lib.strings.fileContents allows a relative path input and works with bootstrapping a new system _and_ rebuilding a running system.

Would be a nice improvement if passwordFile could both accept a full path and relative path I think!

**Comment by OPNA2608** [comment]
> When bootstrapping a new system, everything is mounted into /mnt so I have to deal with the /mnt prefix. Using full path it means something like /mnt/path/to/password.txt.

Have you tried it without the /mnt prefix? [I also use `passwordFile`](https://github.com/OPNA2608/nixos-config/blob/e46b379167080cf43527469e8d869714ee06f8b0/users/puna.nix#L9) and bootstrapping new systems works fine for me with that.

**Comment by kciredor** [comment]
Yes thanks @OPNA2608, passwordFile =  "/etc/nixos/secrets/password-user.txt" does not work during install for me. It results in the '!' "hash" in /etc/shadow.

The file /etc/nixos/secrets/password-user.txt is actually a symlink to /mnt/home/username/blah/password-user.txt during install (and to /home/username/blah/password-user.txt after install). Did not think that would matter. To ensure it does or does not, I've used that exact (destination) value with /mnt/ prefix for passwordFile which does not work either.

The value of /home/username/blah/password-user.txt, omitting /mnt, does finally work with nixos-install!

_Perhaps passwordFile does not follow symlinks during nixos-install, and does follow them during nixos-rebuild?_

**Comment by Infinisil** [comment]
Quoting the description of the [`passwordFile`](https://search.nixos.org/options?channel=21.11&show=users.users.%3Cname%3E.passwordFile&from=0&size=50&sort=relevance&type=packages&query=passwordFile) option:

> The password file is read on each system activation.

And system activation happens when you `nixos-install`, `nixos-rebuild` _and_ when you boot the machine.

**Comment by PgBiel** [comment]
For future readers, I was facing this problem while configuring an impermanence setup (with https://github.com/nix-community/impermanence), as I was trying to persist the hashed password file but failing, even though the file was being bind mounted (and/or symlinked). Thanks to the information in this thread, I figured out that I had to add `fileSystems."/path/to/password/file".neededForBoot = true;` so that the bind mount occurred at the right time, such that the password file was properly picked up, and I could then log in.

**Comment by 5donuts** [comment]
I realize this is a necrobump, but I've been trying to work through a similar problem and came across this issue.
When I create a password hash with `mkpasswd -m sha-512` and use that with `hashedPasswordFile`, I'm not able to login after provisioning a fresh system.

However, when I supply the exact same password hash to the `hashedPassword` option, then I _can_ login.
I suspect this is because my default text editor saves a trailing `\n` to text files.

So, despite `wc -l /path/to/my/hash.txt` indicating the file only has one line, that trailing newline must matter to the `hashedPasswordFile` option.

Folks running into this problem should double-check that the text file containing the password hash is indeed only one line, as noted by the [`hashedPasswordFile` description](https://search.nixos.org/options?channel=unstable&show=users.extraUsers.%3Cname%3E.hashedPasswordFile&from=0&size=50&sort=relevance&type=packages&query=hashedPasswordFile).
